flang/include/flang/Optimizer/Dialect/FIROps.td

```diff
@@ -1133,15 +1134,11 @@ def fir_EmboxOp : fir_Op<"embox", [NoSideEffect, AttrSizedOperandSegments]> {
         if (lenPs != rt.getNumLenParams())
           return emitOpError("number of LEN params does not correspond"
                              " to the !fir.type type");
+      } else if (auto strTy = fir::unwrap_char(eleTy)) {
+        if (strTy.getLen() != fir::CharacterType::unknownLen())
+          return emitOpError("CHARACTER already has static LEN");
       } else {
-        if (auto strTy = eleTy.dyn_cast<fir::SequenceType>()) {
-          if (!strTy.getEleTy().isa<fir::CharacterType>())
-            return emitOpError("LEN parameters require CHARACTER type");
-          if (strTy.getShape()[0] != fir::SequenceType::getUnknownExtent())
-            return emitOpError("CHARACTER already has static LEN");
-        } else {
-          return emitOpError("LEN parameters require !fir.type type");
-        }
+          return emitOpError("LEN parameters require CHARACTER or derived type");
```

Comment:
Doesn't `unwrap_char` always cause CHARACTER variables to take the earlier alternative?

Reply:
It will, but I am not following the question. I want the character to take the earlier alternative since I am complaining this is not a character here.

Reply:
OK, the change to the error message mentions CHARACTER now, but I don't think CHARACTER gets to this alternative. So the error message seems like it might be confusing to a user.

```mlir
... !fir.type<t> ...
```
error: LEN parameters require CHARACTER ...

Reply:
I meant the message to complain that the user provided a type that is not a character or a derived type, and yet that the user provided length parameters. How would you phrase that in a clearer way ?